File: 5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        res = ""

        def expand(l, r):
            while l >= 0 and r <= len(s) - 1 and s[l] == s[r]:
                l -= 1
                r += 1
            return s[l + 1 : r]

        for i in range(len(s)):
            tmp1 = expand(i, i)
            tmp2 = expand(i, i + 1)

            

            if len(tmp1) > len(res):
                res = tmp1
            
            if len(tmp2) > len(res):
                res = tmp2
            
        return res
# @lc code=end

File: 5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        String res = "";

        for (int i = 0; i < s.length(); i++) {
            String tmp1 = this.expand(i, i, s);
            
            if (tmp1.length() > res.length()) {
                res = tmp1;
            }

            String tmp2 = this.expand(i, i + 1, s);

            if (tmp2.length() > res.length()) {
                res = tmp2;
            }
        }

        return res;

    }

    public String expand(int l, int r, String s) {
        while (l >= 0 && r <= s.length() - 1 && s.charAt(l) == s.charAt(r)) {
            l--;
            r++;
        }

        return s.substring(l + 1, r);
    }
}
// @lc code=end

